src/System.Text.Primitives/System/Text/Encoders/Utf8.cs

```diff
@@ -768,30 +496,24 @@ public static OperationStatus FromUtf32(ReadOnlySpan<byte> source, Span<byte> de
             /// <returns>A <see cref="OperationStatus"/> value representing the expected state of the conversion.</returns>
             public static OperationStatus ToUtf32Length(ReadOnlySpan<byte> source, out int bytesNeeded)
             {
-                bytesNeeded = 0;
+                int indexOfFirstInvalidSequence = Utf8Util.GetIndexOfFirstInvalidUtf8Sequence(source, out int scalarCount, out _);
+                bytesNeeded = checked(scalarCount * 4 /* UTF-32 code units are 4 bytes in length */);
```

Comment:
Should we return bytesNeeded = 0 for InvalidData?

Reply:
Fine by me. I'll also change it to return `InvalidData` even if the data is incomplete (implicit _isFinalBlock=true_). Since there's no out parameter that corresponds to how much data was consumed, the caller cannot recover from an "incomplete data" return code, so it should be treated identically to error.